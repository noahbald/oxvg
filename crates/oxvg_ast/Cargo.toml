[package]
name = "oxvg_ast"
version = "0.0.1-beta.1"
edition = "2021"
license.workspace = true
repository.workspace = true
authors.workspace = true
description = "Trait representation for concrete ASTs"

[lints]
workspace = true

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html
[features]
default = ["markup5ever", "roxmltree", "visitor"]
markup5ever = ["dep:rcdom", "dep:xml5ever"]
roxmltree = ["dep:roxmltree"]
serialize = []
selectors = ["dep:selectors", "dep:cssparser", "dep:precomputed-hash"]
visitor = ["dep:bitflags"]
style = ["dep:lightningcss", "dep:cssparser_lightningcss", "dep:smallvec"]

[dependencies]
anyhow = { workspace = true, optional = true }
bitflags = { workspace = true, optional = true }
cfg-if = { workspace = true }
cssparser = { workspace = true, optional = true }
cssparser_lightningcss = { package = "cssparser", version = "0.33", optional = true }
derive_more = { workspace = true, features = ["debug"] }
itertools = { workspace = true }
lightningcss = { workspace = true, optional = true }
log = { workspace = true }
markup5ever = { workspace = true }
precomputed-hash = { workspace = true, optional = true }
rcdom = { workspace = true, optional = true }
roxmltree = { workspace = true, optional = true }
# ensure track_alloc_size for wasm compat
servo_arc = { version = "0.4", features = ["track_alloc_size"] }
selectors = { workspace = true, optional = true }
smallvec = { version = "1.13.2", optional = true }
string_cache = { workspace = true }
tendril = { workspace = true }
typed-arena = { workspace = true }
xml5ever = { workspace = true, optional = true }
