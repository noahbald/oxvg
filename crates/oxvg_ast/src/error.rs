//! Error types.

use std::fmt::Display;

/// A parser error
#[derive(Debug)]
pub enum ParseErrorKind<'i> {
    /// Lightningcss failed to parse
    CSSParserError(lightningcss::error::ParserError<'i>),
    /// A fundamental parsing step failed
    Basic(cssparser_lightningcss::BasicParseError<'i>),
    /// A clock value didn't fit the expected range
    InvalidClockValue,
    /// A unexpected set of paint steps were given
    InvalidPaintOrder,
    /// A begin-end syncbase value is missing an id
    MissingSyncbaseId,
}

/// Parse errors that can be encoutered by parsing
pub type ParseError<'input> = cssparser_lightningcss::ParseError<'input, ParseErrorKind<'input>>;

impl<'input> ParseErrorKind<'input> {
    /// Maps a lightnincss error to `Self`
    pub fn from_css(
        error: cssparser_lightningcss::ParseError<'input, lightningcss::error::ParserError<'input>>,
    ) -> cssparser_lightningcss::ParseError<'input, ParseErrorKind<'input>> {
        match error.kind {
            cssparser_lightningcss::ParseErrorKind::Basic(e) => {
                cssparser_lightningcss::ParseError {
                    kind: cssparser_lightningcss::ParseErrorKind::Basic(e),
                    location: error.location,
                }
            }
            cssparser_lightningcss::ParseErrorKind::Custom(e) => {
                cssparser_lightningcss::ParseError {
                    kind: cssparser_lightningcss::ParseErrorKind::Custom(
                        ParseErrorKind::CSSParserError(e),
                    ),
                    location: error.location,
                }
            }
        }
    }

    /// Maps a basic error to `Self`
    pub fn from_basic(
        error: cssparser_lightningcss::BasicParseError<'input>,
    ) -> cssparser_lightningcss::ParseError<'input, Self> {
        cssparser_lightningcss::ParseError {
            kind: cssparser_lightningcss::ParseErrorKind::Basic(error.kind),
            location: error.location,
        }
    }

    /// Maps a parser error to `Self`
    pub fn from_parser(
        error: cssparser_lightningcss::ParseError<'input, cssparser_lightningcss::BasicParseError>,
    ) -> cssparser_lightningcss::ParseError<'input, Self> {
        match error.kind {
            cssparser_lightningcss::ParseErrorKind::Basic(e) => {
                cssparser_lightningcss::ParseError {
                    kind: cssparser_lightningcss::ParseErrorKind::Basic(e),
                    location: error.location,
                }
            }
            cssparser_lightningcss::ParseErrorKind::Custom(e) => {
                cssparser_lightningcss::ParseError {
                    kind: cssparser_lightningcss::ParseErrorKind::Custom(ParseErrorKind::Basic(e)),
                    location: error.location,
                }
            }
        }
    }
}

/// A printer error
pub type PrinterError = lightningcss::error::PrinterError;

/// An error while serializing a document.
#[derive(Debug)]
pub enum XmlWriterError {
    /// An error while running an io operation.
    IO(std::io::Error),
    /// An error while formatting
    FMT(std::fmt::Error),
    /// An error while flushing buffer.
    BufWriter(std::io::IntoInnerError<std::io::BufWriter<Vec<u8>>>),
    /// An error after writing to string.
    UTF8(std::string::FromUtf8Error),
    /// An error while serializing an attribute
    PrinterError(PrinterError),
    /// Did not have opening element name when closing element.
    ClosedUnopenedElement,
    /// Attempted to write attribute before `start_element()` or after `close_element()`.
    AttributeWrittenBeforeElement,
    /// Declaration was already written.
    DeclarationAlreadyWritten,
    /// Attempts to write text before `start_element()`.
    TextBeforeElement,
    /// Attempts to write CDATA with `]]>` in the content.
    BadCDATA,
}
impl std::fmt::Display for XmlWriterError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            Self::IO(err) => err.fmt(f),
            Self::FMT(err) => err.fmt(f),
            Self::BufWriter(err) => err.fmt(f),
            Self::UTF8(err) => err.fmt(f),
            Self::PrinterError(err) => err.fmt(f),
            Self::ClosedUnopenedElement => {
                "Did not have opening element name when closing element.".fmt(f)
            }
            Self::AttributeWrittenBeforeElement => {
                "Attempted to write attribute before `start_element()` or after `close_element()`."
                    .fmt(f)
            }
            Self::TextBeforeElement => "Attempts to write text before `start_element()`.".fmt(f),
            Self::BadCDATA => "Attempts to write CDATA with `]]>` in the content.".fmt(f),
            Self::DeclarationAlreadyWritten => "Declaration was already written.".fmt(f),
        }
    }
}
impl std::error::Error for XmlWriterError {}

/// An error while gathering computed styles
#[derive(Debug, Clone)]
pub enum ComputedStylesError<'input> {
    /// A selector was found that may affect validity of computed styles
    BadSelector(lightningcss::selector::SelectorList<'input>),
}
impl Display for ComputedStylesError<'_> {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            Self::BadSelector(selector) => {
                f.write_str("Found an invalid selector while processing: ")?;
                selector.fmt(f)
            }
        }
    }
}
impl std::error::Error for ComputedStylesError<'_> {}
