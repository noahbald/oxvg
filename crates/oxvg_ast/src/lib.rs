#[macro_use]
extern crate bitflags;

#[cfg(feature = "style")]
#[macro_use]
extern crate smallvec;

pub mod atom;
pub mod attribute;
pub mod class_list;
pub mod document;
pub mod element;
pub mod implementations;
pub mod name;
pub mod node;

#[cfg(feature = "visitor")]
pub mod visitor;

#[cfg(feature = "parse")]
pub mod parse;

#[cfg(feature = "serialize")]
pub mod serialize;

#[cfg(feature = "selectors")]
pub mod selectors;

#[cfg(feature = "style")]
pub mod style;
