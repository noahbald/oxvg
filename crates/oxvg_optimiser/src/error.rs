use std::{fmt::Display, sync::mpsc::RecvTimeoutError};

use oxvg_ast::{atom::Atom, error::ComputedStylesError};

#[derive(Debug, Clone)]
pub enum PrecheckError<'input> {
    ScriptingNotSupported,
    AnimationNotSupported,
    ConditionalProcessingNotSupported,
    ReferencesExternalXLink(Atom<'input>),
}
impl Display for PrecheckError<'_> {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            Self::ScriptingNotSupported => f.write_str("scripting is not supported"),
            Self::AnimationNotSupported => f.write_str("animation is not supported"),
            Self::ConditionalProcessingNotSupported => {
                f.write_str("conditional processing attributes is not supported")
            }
            Self::ReferencesExternalXLink(xlink) => f.write_fmt(format_args!("the `xlink:href` attribute is referencing an external object '{xlink}' which is not supported"))
        }
    }
}
impl std::error::Error for PrecheckError<'_> {}

#[derive(Debug, Clone)]
#[cfg(feature = "napi")]
pub enum PrefixGeneratorError {
    ThreadTimeoutError(RecvTimeoutError),
    ThreadUnknownError,
    NapiError(napi::Status),
    ClientError(String),
}

#[derive(Debug, Clone)]
pub enum JobsError<'input> {
    ComputedStylesError(ComputedStylesError<'input>),
    CleanupValuesPrecision(u8),
    Precheck(PrecheckError<'input>),
    #[cfg(feature = "napi")]
    PrefixGenerator(PrefixGeneratorError),
    InvalidUserSelector(String),
    InvalidUserRegex(regex::Error),
    #[deprecated]
    Generic(String),
}
impl Display for JobsError<'_> {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            Self::ComputedStylesError(e) => e.fmt(f),
            Self::CleanupValuesPrecision(n) => f.write_fmt(format_args!(
                "The float-precision `{n}` is larger than the maximum of 5"
            )),
            Self::Precheck(e) => e.fmt(f),
            #[cfg(feature = "napi")]
            Self::PrefixGenerator(_) => f.write_str("Prefix generate failed to close threads"),
            Self::InvalidUserSelector(e) => {
                f.write_fmt(format_args!("Invalid selector in configuration: {e}"))
            }
            Self::InvalidUserRegex(e) => e.fmt(f),
            Self::Generic(e) => e.fmt(f),
        }
    }
}
impl std::error::Error for JobsError<'_> {}
