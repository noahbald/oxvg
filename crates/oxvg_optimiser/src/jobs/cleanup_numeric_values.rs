use oxvg_ast::{
    element::Element,
    visitor::{Context, Visitor},
};
use serde::{Deserialize, Serialize};

#[cfg(feature = "wasm")]
use tsify::Tsify;

use crate::error::JobsError;

#[cfg_attr(feature = "wasm", derive(Tsify))]
#[cfg_attr(feature = "napi", napi(object))]
#[derive(Deserialize, Serialize, Debug, Clone)]
#[serde(rename_all = "camelCase")]
/// Rounds number and removes default `px` unit in attributes specified with a number number.
///
/// # Correctness
///
/// Rounding errors may cause slight changes in visual appearance.
///
/// # Errors
///
/// When a float-precision greater than the maximum is given.
pub struct CleanupNumericValues {
    #[serde(default = "default_float_precision")]
    // WARN: Lightningcss will round values to 5 decimal places
    /// Number of decimal places to round floating point numbers to, to a maximum of 5.
    pub float_precision: u8,
    #[serde(default = "default_leading_zero")]
    #[deprecated(note = "This option has no effect; leading zeroes are always removed")]
    /// Whether to trim leading zeros.
    pub leading_zero: bool,
    #[serde(default = "default_default_px")]
    /// Whether to remove `px` from a number's unit.
    pub default_px: bool,
    #[serde(default = "default_convert_to_px")]
    /// Whether to convert absolute units like `cm` and `in` to `px`.
    pub convert_to_px: bool,
}

impl Default for CleanupNumericValues {
    fn default() -> Self {
        CleanupNumericValues {
            float_precision: default_float_precision(),
            leading_zero: default_leading_zero(),
            default_px: default_default_px(),
            convert_to_px: default_convert_to_px(),
        }
    }
}

impl<'input, 'arena> Visitor<'input, 'arena> for CleanupNumericValues {
    type Error = JobsError<'input>;

    fn document(
        &self,
        _document: &Element<'input, 'arena>,
        _context: &Context<'input, 'arena, '_>,
    ) -> Result<(), Self::Error> {
        if self.float_precision > 5 {
            Err(JobsError::CleanupValuesPrecision(self.float_precision))
        } else {
            Ok(())
        }
    }

    fn element(
        &self,
        element: &Element<'input, 'arena>,
        _context: &mut Context<'input, 'arena, '_>,
    ) -> Result<(), Self::Error> {
        element.attributes().into_iter_mut().for_each(|mut attr| {
            attr.value_mut()
                .round(self.float_precision as f32, self.convert_to_px, false);
        });
        Ok(())
    }
}

const fn default_float_precision() -> u8 {
    3
}
const fn default_leading_zero() -> bool {
    true
}
const fn default_default_px() -> bool {
    true
}
const fn default_convert_to_px() -> bool {
    true
}

#[test]
fn cleanup_numeric_values() -> anyhow::Result<()> {
    use crate::test_config;

    insta::assert_snapshot!(test_config(
        r#"{ "cleanupNumericValues": {} }"#,
        Some(
            r#"<svg xmlns="http://www.w3.org/2000/svg" viewBox="20.000001 -19.99999 17.123456 70.708090" width="50.12356%" height="20px" x=".2655" y="-.2346">
    <!-- Should round values, maintaining non-numerical values -->
    <rect width="1in" height="12pt"/>
</svg>"#
        )
    )?);

    insta::assert_snapshot!(test_config(
        r#"{ "cleanupNumericValues": {} }"#,
        Some(
            r#"<svg xmlns="http://www.w3.org/2000/svg" width="20" height="20" viewBox="0, 0, 20, 20">
    <!-- Should round values, maintaining non-numerical values -->
    <rect width="20" height="20" fill="rgba(255,255,255,.85)" rx="20"/>
</svg>"#
        )
    )?);

    insta::assert_snapshot!(test_config(
        r#"{ "cleanupNumericValues": {} }"#,
        Some(
            r#"<svg xmlns="http://www.w3.org/2000/svg" viewBox=" 0 0      150 100 ">
    <!-- Should remove unnecessary whitespace from `viewBox -->
</svg>"#
        )
    )?);

    insta::assert_snapshot!(test_config(
        r#"{ "cleanupNumericValues": { "floatPrecision": 0 } }"#,
        Some(
            r#"<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 1052.4 744.1">
    <!-- Should round to zero decimal places -->
</svg>"#
        )
    )?);

    Ok(())
}
