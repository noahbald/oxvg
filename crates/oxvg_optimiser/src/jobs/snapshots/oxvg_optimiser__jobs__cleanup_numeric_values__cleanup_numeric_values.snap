---
source: crates/oxvg_optimiser/src/jobs/cleanup_numeric_values.rs
expression: "test_config(r#\"{ \"cleanupNumericValues\": {} }\"#,\nSome(r#\"<svg xmlns=\"http://www.w3.org/2000/svg\" viewBox=\"20.000001 -19.99999 17.123456 70.708090\" width=\"50.12356%\" height=\"20px\" x=\".2655\" y=\"-.2346\">\n    <!-- Should round values, maintaining non-numerical values -->\n    <rect width=\"1in\" height=\"12pt\"/>\n</svg>\"#))?"
---
<svg xmlns="http://www.w3.org/2000/svg" viewBox="20 -20 17.123 70.708" width="50.124%" height="20" x=".266" y="-.235">
    <!-- Should round values, maintaining non-numerical values -->
    <rect width="96" height="16"/>
</svg>
